### PR TITLE
Window leak fix

### LINKED ISSUE #2913: Directional Windows + Atmos
When a directional (single-pane, not full) window is secured to a floor, then the tile under it is removed, then the plating has a reinforced floor (metal rods instead of floor tiles) placed under it, it loses its airtight quality. It can be fixed by using a screwdriver twice. I don't know if this is a feature or a bug, but it is rather misleading - I wouldn't have realized it had the floor under me not started turning icy.

### PR DESCRIPTION
**Pull request by abyll**
## What Does This PR Do
-     Finishes the windows fixes started by Kiyahitayika - credit to him for doing most of the work.
-     Fixes the window atmos leak issue when you move fulltile windows
-     Fixes the problem with atmos leaking under windows when you pry up their tile, or freshly build it. - resolves #2913 and #15768
-     Deletes the fulltile variable, replaces the very few instances where this was referenced with is_fulltile()
-     Makes window/full a proper subtype with overrides (except for CanAtmosPass(), my arch nemesis)
-     Tidies up the Clockwork window types

## Why It's Good For The Game
Atmos blows, this makes it blow less.

## Changelog
:cl:
fix: Atmos permeability is now updated at screwdriver anchoring/unanchoring step of all windows. 
fix: Turf update (E.G. removing floor tile) doesn't pull gas from adjacent tiles if the adjacent tile doesn't allow atmos through.
/:cl:

### CODE REVIEW

File: code/__DEFINES/is_helpers.dm [hunk 1]
```diff
@@ -28,6 +28,8 @@
 
 #define is_pen(W) (istype(W, /obj/item/pen))
 
+#define is_fulltile(W) (istype(W, /obj/structure/window/full))
```
Comment (dearmochi) [inline code/__DEFINES/is_helpers.dm:31]:
```suggestion
#define isfulltile(W) (istype(W, /obj/structure/window/full))
```

File: code/__HELPERS/unsorted.dm [hunk 2]
```diff
@@ -156,16 +156,17 @@ Turf and target are seperate in case you want to teleport some distance from a t
 	for(var/obj/structure/window/D in loc)
 		if(!D.density)
 			continue
-		if(D.fulltile)
-			return 1
+		if(is_fulltile(D))
```
Comment (dearmochi) [inline code/__HELPERS/unsorted.dm:159]:
```suggestion
		if(isfulltile(D))
```

File: code/game/gamemodes/miniantags/revenant/revenant_abilities.dm [hunk 3]
```diff
@@ -258,7 +258,7 @@
 					T.ChangeTurf(/turf/simulated/wall/r_wall/rust)
 				for(var/obj/structure/window/window in T.contents)
 					window.take_damage(rand(30,80))
-					if(window && window.fulltile)
+					if(window && is_fulltile(window))
```
Comment (dearmochi) [inline code/game/gamemodes/miniantags/revenant/revenant_abilities.dm:261]:
```suggestion
					if(window && isfulltile(window))
```

File: code/game/objects/structures/window.dm [hunk 4]
```diff
@@ -44,26 +45,21 @@
 			. += "<span class='notice'>The window is <b>screwed</b> to the floor.</span>"
 		else
 			. += "<span class='notice'>The window is <i>unscrewed</i> from the floor, and could be deconstructed by <b>wrenching</b>.</span>"
-	if(!anchored && !fulltile)
+	if(!anchored && !is_fulltile(src))
```
Comment (dearmochi) [inline code/game/objects/structures/window.dm:48]:
```suggestion
	if(!anchored && !isfulltile(src))
```

File: code/game/objects/structures/window.dm [hunk 5]
```diff
@@ -44,26 +45,21 @@
...
 		. += "<span class='notice'>Alt-click to rotate it.</span>"
 
 /obj/structure/window/Initialize(mapload, direct)
 	. = ..()
 
-	if(direct)
+	if(direct && !is_fulltile(src))
```
Comment (dearmochi) [inline code/game/objects/structures/window.dm:54]:
```suggestion
	if(direct && !isfulltile(src))
```

File: code/game/objects/structures/window.dm [hunk 6]
```diff
@@ -88,16 +81,11 @@
 		deconstruct(FALSE)
 
 /obj/structure/window/setDir(direct)
-	if(!fulltile)
-		..()
-	else
-		..(FULLTILE_WINDOW_DIR)
+	..()
```
Comment (dearmochi) [inline code/game/objects/structures/window.dm:84]: The proc override can just be removed as it doesn't do anything new

File: code/game/objects/structures/window.dm [hunk 7]
```diff
@@ -302,18 +290,23 @@
 /obj/structure/window/proc/check_state_and_anchored(checked_state, checked_anchored)
 	return check_state(checked_state) && check_anchored(checked_anchored)
 
+/obj/structure/window/proc/Secured()
```
Comment (dearmochi) [inline code/game/objects/structures/window.dm:293]:
```suggestion
/obj/structure/window/proc/secured()
```

File: code/game/objects/structures/window.dm [hunk 8]
```diff
@@ -436,9 +427,19 @@
 	move_update_air(T)
 
 /obj/structure/window/CanAtmosPass(turf/T)
-	if(!anchored || !density)
+	if(!istype(T))
+		return FALSE
+	if(!density || get_turf(loc) == T)
 		return TRUE
-	return !(FULLTILE_WINDOW_DIR == dir || dir == get_dir(loc, T))
+	if(is_fulltile(src))	// This won't work with an override on /window/full
```
Comment (dearmochi) [inline code/game/objects/structures/window.dm:434]:
```suggestion
	if(isfulltile(src))	// This won't work with an override on /window/full
```

File: code/game/objects/structures/window.dm [hunk 9]
```diff
@@ -436,9 +427,19 @@
...
+		return !Secured()
```
Comment (dearmochi) [inline code/game/objects/structures/window.dm:435]:
```suggestion
		return !secured()
```

File: code/game/objects/structures/window.dm [hunk 10]
```diff
@@ -436,9 +427,19 @@
...
+	else					// Welcome to the realm of Directional Windows
+		if(!(get_dir(loc, T) in GLOB.cardinal))
+			return FALSE	// Directional windows only face NORTH, SOUTH, EAST, WEST
+		if(!Secured())
```
Comment (dearmochi) [inline code/game/objects/structures/window.dm:439]:
```suggestion
		if(!secured())
```

File: code/game/objects/structures/window.dm [hunk 11]
```diff
@@ -447,31 +448,18 @@
...
 	if(exposed_temperature > (T0C + heat_resistance))
 		take_damage(round(exposed_volume / 100), BURN, 0, 0)
 
 /obj/structure/window/GetExplosionBlock()
-	return reinf && fulltile ? real_explosion_block : 0
+	return 0
 
 /obj/structure/window/basic
-	desc = "It looks thin and flimsy. A few knocks with... anything, really should shatter it."
+	desc = "It looks thin and flimsy. A few knocks with ... anything, really, should shatter it."
```
Comment (dearmochi) [inline code/game/objects/structures/window.dm:462]:
```suggestion
	desc = "It looks thin and flimsy. A few knocks with... anything, really, should shatter it."
```

File: code/game/objects/structures/window.dm [hunk 12]
```diff
@@ -579,17 +567,85 @@
...
+
+/obj/structure/window/full/Initialize(mapload, direct)
+	. = ..()
+	if(decon_speed == null)
+		decon_speed = 2 SECONDS
+
+/obj/structure/window/full/can_be_reached(mob/user)
+	return TRUE
+
+/obj/structure/window/full/CanPass(atom/movable/mover, turf/target, height=0)
```
Comment (dearmochi) [inline code/game/objects/structures/window.dm:594]:
```suggestion
/obj/structure/window/full/CanPass(atom/movable/mover, turf/target, height = 0)
```

File: code/game/objects/structures/window.dm [hunk 13]
```diff
@@ -579,17 +567,85 @@
...
+		var/ratio = obj_integrity / max_integrity
+		ratio = CEILING(ratio * 4, 1) * 25
+
+		if(smooth)
+			queue_smooth(src)
+
+		cut_overlay(crack_overlay)
+		if(ratio > 75)
+			return
+		crack_overlay = mutable_appearance('icons/obj/structures.dmi', "damage[ratio]", -(layer+0.1))
```
Comment (dearmochi) [inline code/game/objects/structures/window.dm:619]:
```suggestion
		crack_overlay = mutable_appearance('icons/obj/structures.dmi', "damage[ratio]", -(layer + 0.1))
```

File: code/game/objects/structures/window.dm [hunk 14]
```diff
@@ -579,17 +567,85 @@
...
+/obj/structure/window/full/plasmareinforced/BlockSuperconductivity()
+	return TRUE
+
+/obj/structure/window/full/plasmareinforced/temperature_expose(datum/gas_mixture/air, exposed_temperature, exposed_volume)
+	return
+
+//Types
 /obj/structure/window/full/basic
-	desc = "It looks thin and flimsy. A few knocks with... anything, really should shatter it."
+	desc = "It looks thin and flimsy. A few knocks with ... anything, really, should shatter it."
```
Comment (dearmochi) [inline code/game/objects/structures/window.dm:648]:
```suggestion
	desc = "It looks thin and flimsy. A few knocks with... anything, really, should shatter it."
```

File: code/modules/admin/verbs/mapping.dm [hunk 15]
```diff
@@ -88,7 +88,7 @@ GLOBAL_VAR_INIT(intercom_range_display_status, 0)
 			if(!(locate(/obj/structure/grille,T)))
 				var/window_check = 0
 				for(var/obj/structure/window/W in T)
-					if(W.dir == turn(C1.dir,180) || W.fulltile)
+					if(W.dir == turn(C1.dir,180) || is_fulltile(W))
```
Comment (dearmochi) [inline code/modules/admin/verbs/mapping.dm:91]:
```suggestion
					if(W.dir == turn(C1.dir, 180) || isfulltile(W))
```

File: code/modules/mob/living/living.dm [hunk 16]
```diff
@@ -180,12 +180,10 @@
 	if((AM.anchored && !push_anchored) || (force < (AM.move_resist * MOVE_FORCE_PUSH_RATIO)))
 		now_pushing = FALSE
 		return
-	if(istype(AM, /obj/structure/window))
-		var/obj/structure/window/W = AM
-		if(W.fulltile)
-			for(var/obj/structure/window/win in get_step(W,t))
-				now_pushing = FALSE
-				return
+	if(is_fulltile(AM))
```
Comment (dearmochi) [inline code/modules/mob/living/living.dm:183]:
```suggestion
	if(isfulltile(AM))
```

File: code/modules/mob/living/simple_animal/hostile/megafauna/swarmer.dm [hunk 17]
```diff
@@ -188,8 +188,9 @@ GLOBAL_LIST_INIT(AISwarmerCapsByType, list(/mob/living/simple_animal/hostile/swa
 	//SPECIFIC CASES:
 	//Smash fulltile windows before grilles
 	if(istype(the_target, /obj/structure/grille))
-		for(var/obj/structure/window/rogueWindow in get_turf(the_target))
-			if(rogueWindow.fulltile) //done this way because the subtypes are weird.
+		for(var/obj/structure/window/W in get_turf(the_target))
+			if(is_fulltile(W))
```
Comment (dearmochi) [inline code/modules/mob/living/simple_animal/hostile/megafauna/swarmer.dm:192]:
```suggestion
			if(isfulltile(W))
```

**Review by dearmochi — CHANGES_REQUESTED** [review]

**Comment by SteelSlayer** [comment]
Does your PR fix issue #15768?

**Comment by abyll** [comment]
> 
> 
> Does your PR fix issue #15768?

It should, all variations of reinforced windows, fulltile and regular, seem to be working as intended now.

**Comment by variableundefined** [comment]
Conflicted since 2 months ago. Open new PR when deconflicted.

**Pull request closed without merging**